File: app/src/main/java/com/mouredev/weeklychallenge2022/Challenge24.kt
package com.mouredev.weeklychallenge2022

/*
 * Reto #24
 * ITERATION MASTER
 * Fecha publicación enunciado: 13/06/22
 * Fecha publicación resolución: 20/06/22
 * Dificultad: FÁCIL
 *
 * Enunciado: Quiero contar del 1 al 100 de uno en uno (imprimiendo cada uno). ¿De cuántas maneras eres capaz de hacerlo? Crea el código para cada una de ellas.
 *
 * Información adicional:
 * - Usa el canal de nuestro discord (https://mouredev.com/discord) "🔁reto-semanal" para preguntas, dudas o prestar ayuda a la comunidad.
 * - Puedes hacer un Fork del repo y una Pull Request al repo original para que veamos tu solución aportada.
 * - Revisaré el ejercicio en directo desde Twitch el lunes siguiente al de su publicación.
 * - Subiré una posible solución al ejercicio el lunes siguiente al de su publicación.
 *
 */

fun main() {

    // 1

    println("**** 1 ****")

    for (index in (1..100)) {
        println(index)
    }

    // 2

    println("**** 2 ****")

    (1..100).forEach { index ->
        println(index)
    }

    // 3

    println("**** 3 ****")

    var whileIndex = 1

    while (whileIndex <= 100) {
        println(whileIndex)
        whileIndex += 1
    }

    // 4

    println("**** 4 ****")

    whileIndex = 1

    do {
        println(whileIndex)
        whileIndex += 1
    } while (whileIndex <= 100)

    // 5

    println("**** 5 ****")

    fun print100(index: Int) {
        if (index <= 100) {
            println(index)
            print100(index + 1)
        }
    }

    print100(1)

    // 6

    println("**** 6 ****")

    println((1..100).filter {
        true
    })

    // 7

    println("**** 7 ****")

    println((1..100).map { it })
}

File: app/src/main/java/com/mouredev/weeklychallenge2022/Challenge24.dart
/*
 * Reto #24
 * ITERATION MASTER
 * Fecha publicación enunciado: 13/06/22
 * Fecha publicación resolución: 20/06/22
 * Dificultad: FÁCIL
 *
 * Enunciado: Quiero contar del 1 al 100 de uno en uno (imprimiendo cada uno). ¿De cuántas maneras eres capaz de hacerlo? Crea el código para cada una de ellas.
 *
 * Información adicional:
 * - Usa el canal de nuestro discord (https://mouredev.com/discord) "🔁reto-semanal" para preguntas, dudas o prestar ayuda a la comunidad.
 * - Puedes hacer un Fork del repo y una Pull Request al repo original para que veamos tu solución aportada.
 * - Revisaré el ejercicio en directo desde Twitch el lunes siguiente al de su publicación.
 * - Subiré una posible solución al ejercicio el lunes siguiente al de su publicación.
 *
 */

void main(List<String> args) {
  //Contar con bucle FOR--------------------------------------------------------
  void countFor(int from, int to) {
    for (; from <= to; print(from++));
  }

  //Contar con bucle WHILE------------------------------------------------------
  void countWhile(int from, int to) {
    while (from <= to) print(from++);
  }

  //Contar con RECURSION--------------------------------------------------------
  void countRecursion(int from, int to) {
    if (to > from) countRecursion(from, to - 1);
    print(to);
  }

  //Contar con bucle DO..WHILE--------------------------------------------------
  void countDoWhile(int from, int to) {
    do {
      print(from++);
    } while (from <= to);
  }

  //Contar usando operaciones de BIT--------------------------------------------
  void countBitOp(int from, int to) {
    while (from <= to) {
      print(from);
      int temp = 1;
      // mientras el OR sea menor o igual al actual
      //Ej1: 00101 (5) | 00001 (1) -> 00101 (5) <= 00101 (5)
      //Ej2: 00101 (5) | 00011 (3) -> 00111 (7) <= 00101 (5)
      while (from | temp <= from) {
        temp <<= 1; // Ej: 00001 (1) -> 00010 (2)
        temp |= 1; // Ej: 00010 (2) -> 00011 (3)
      }
      from ^= temp; // Ej: 00101 (5) ^ 00011 (3) -> 00110 (6)
    }
  }

  //Contar usando una lista con FOR-EACH----------------------------------------
  void countForEach(int from, int to) {
    List.generate(to, (index) => from + index)
      ..forEach((element) {
        print(element);
      });
  }

  //MAIN------------------------------------------------------------------------
  int from = 1;
  int to = 100;
  if (from > to) {
    int temp = from;
    from = to;
    to = temp;
  }

  // countFor(from, to);
  // countWhile(from, to);
  // countRecursion(from, to);
  // countDoWhile(from, to);
  // countBitOp(from, to);
  // countForEach(from, to);
}
